src/froll.c

```diff
@@ -1,52 +1,75 @@
-  if (nx < k) {                                                 // if window width bigger than input just return vector of fill values
+void frollfun(rollfun_t rfun, unsigned int algo, double *x, uint64_t nx, ans_t *ans, int k, int align, double fill, bool narm, int hasna, bool verbose) {
+  double tic = 0;
+  if (verbose)
+    tic = omp_get_wtime();
+  if (nx < k) {                      // if window width bigger than input just return vector of fill values
     if (verbose)
       snprintf(end(ans->message[0]), 500, _("%s: window width longer than input vector, returning all NA vector\n"), __func__);
     // implicit n_message limit discussed here: https://github.com/Rdatatable/data.table/issues/3423#issuecomment-487722586
-    for (int i=0; i<nx; i++) {
+    for (uint64_t i=0; i<nx; i++) {
       ans->dbl_v[i] = fill;
     }
     return;
   }
-  double tic = 0;
-  if (verbose)
-    tic = omp_get_wtime();
-  if (algo==0) {
-    frollmeanFast(x, nx, ans, k, fill, narm, hasna, verbose);
-  } else if (algo==1) {
-    frollmeanExact(x, nx, ans, k, fill, narm, hasna, verbose);
+  switch (rfun) {
+  case MEAN :
+    if (algo==0) {
+      frollmeanFast(x, nx, ans, k, fill, narm, hasna, verbose);
+    } else if (algo==1) {
+      frollmeanExact(x, nx, ans, k, fill, narm, hasna, verbose);
+    }
+    break;
+  case SUM :
+    if (algo==0) {
+      frollsumFast(x, nx, ans, k, fill, narm, hasna, verbose);
+    } else if (algo==1) {
+      frollsumExact(x, nx, ans, k, fill, narm, hasna, verbose);
+    }
+    break;
+  case MAX :
+    if (algo==0) {
+      frollmaxFast(x, nx, ans, k, fill, narm, hasna, verbose);
+    } else if (algo==1) {
+      frollmaxExact(x, nx, ans, k, fill, narm, hasna, verbose);
+    }
+    break;
+  default:
+    error(_("Internal error: Unknown rfun value in froll: %d"), rfun); // #nocov
   }
-  if (ans->status < 3 && align < 1) {                           // align center or left, only when no errors occurred
-    int k_ = align==-1 ? k-1 : floor(k/2);                      // offset to shift
+  if (align < 1 && ans->status < 3) {
+    int k_ = align==-1 ? k-1 : floor(k/2);       // offset to shift
     if (verbose)
       snprintf(end(ans->message[0]), 500, _("%s: align %d, shift answer by %d\n"), __func__, align, -k_);
     memmove((char *)ans->dbl_v, (char *)ans->dbl_v + (k_*sizeof(double)), (nx-k_)*sizeof(double)); // apply shift to achieve expected align
-    for (uint64_t i=nx-k_; i<nx; i++) {                         // fill from right side
+    for (uint64_t i=nx-k_; i<nx; i++) {          // fill from right side
       ans->dbl_v[i] = fill;
     }
   }
   if (verbose)
-    snprintf(end(ans->message[0]), 500, _("%s: processing algo %u took %.3fs\n"), __func__, algo, omp_get_wtime()-tic);
+    snprintf(end(ans->message[0]), 500, _("%s: processing fun %d algo %u took %.3fs\n"), __func__, rfun, algo, omp_get_wtime()-tic);
```

Comment:
Same as below
`frolladaptivefun: processing fun 2 algo 0 took`
reads quite cryptic

Reply:
let me address it there